### PR TITLE
IR refactor: create ConstraintBuilder

### PR DESCRIPTION
**Pull request by grjte**
This PR is the first in a series addressing #150. This one simply reorganizes the IR crate but does not yet move the functions that build the graph into the `ConstraintBuilder` as mentioned in that issue. That will be addressed in a subsequent PR. There are a number of `TODOs` throughout the code which will be removed by the end of this refactor

The next PRs will:

- [ ] simplify symbol validation
- [ ] refactor the symbol table to contain scoped symbols
- [ ] move the functions that process expressions from the `AlgebraicGraph` to the `ConstraintBuilder`

### CODE REVIEW

File: ir/src/symbol_table/mod.rs [hunk 1]
```diff
@@ -41,36 +48,36 @@ impl Display for IdentifierType {
...
+
+// SYMBOL TABLE
+// ================================================================================================
+
 /// SymbolTable for identifiers to track their types and information and enforce uniqueness of
 /// identifiers.
 #[derive(Default, Debug)]
 pub struct SymbolTable {
     /// Vector in which index is trace segment and value is number of columns in this segment
     segment_widths: Vec<u16>,
```
Comment (Overcastan) [inline ir/src/symbol_table/mod.rs:74]: Looks like this field was also moved to the Declarations struct, should we remove it from here?

**Review by Overcastan — COMMENTED** [review]
Looks good to me, just one question inline

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Review by tohrnii — APPROVED** [review]
Thank you @grjte, looks great!

File: ir/src/symbol_table/mod.rs (hunk 1 above)
Comment (grjte) [reply ir/src/symbol_table/mod.rs:74]: Great catch! Yes, we should remove it from here and move the `set` handling into the declarations as well. Thank you!

Comment (grjte) [reply ir/src/symbol_table/mod.rs]: I've fixed it. Will you please review the update?

**Comment by grjte** [comment]
I'm planning to wait to merge this until we've merged #147, #154, and #172.

**Review by Overcastan — APPROVED** [review]
Looks great, thank you @grjte!

**Pull request merged**